### ISSUE TITLE
DataObjects can declare their own indexing behaviour

### ISSUE THREAD

**Issue by unclecheese** [silverstripe/silverstripe-search-service#6]
As a Silverstripe CMS developer, I want to leverage the power of the ORM to create a highly inclusive set of index fields on my dataobjects so that searches return more accurate results.

## Acceptance Criteria

- [x] Can whitelist both native and relationship fields for indexing
- [x] Can list getters (computed fields) for indexing
- [ ] ~Include all fields on subclasses~ Note: Decided not to do this to avoid inferring field names
- [x] Individual records can opt out of indexing (e.g. ShowInSearch)

**Comment by chillu** [comment]
> Include all fields on subclasses

I think we should allow for selecting fields on subclasses, but not a `*` rule where it includes everything - because of https://github.com/silverstripe/silverstripe-search-service/issues/18.

```yml
Page:
  fields:
    Content
    MyCustomFieldOnSubclass
```

**Comment by dhensby** [comment]
It would be nice if you could index any object regardless of if it is from the ORM, for example by making them comply with an `interface` or trait.

The default behaviour could be supplied as a trait that is easily added to `DataObject`s.

Should be fairly simple by implementing a `SearchServiceIndexable` interface and a `DataObjectIndexable` trait (naming could be anything)

**Comment by unclecheese** [comment]
Yes, this is the intention: https://github.com/silverstripe/silverstripe-search-service/blob/master/src/Interfaces/DocumentInterface.php

And the DataObject implementation: https://github.com/silverstripe/silverstripe-search-service/blob/master/src/DataObject/DataObjectDocument.php

**Comment by unclecheese** [comment]
Removing "index all fields on subclasses" as Ingo and I have discussed the potential security hazards of implicit exposure of fields.

**Comment by chillu** [comment]
> Can whitelist both native and relationship fields for indexing

This doesn't work with subclasses at the moment:

```yml
SilverStripe\SearchService\Service\IndexConfiguration:
  indexes:
    my_index:
      includeClasses:
        Page:
          fields:
            title:
              property: Title
            relation_only_on_subclass:
              property: MyRelation.Title
```

`DataObject->relObject()` throws when it can't find the relationship. I'm not keen to create subclass-specific index definitions, since we should use this layer to communicate the "flattening" which happens in App Search (doesn't know about subclasses). We could catch the `LogicException` but that's messy. As a workaround, we could check if the first segment of the dot notation name exists on the object?

```diff
diff --git a/src/DataObject/DataObjectDocument.php b/src/DataObject/DataObjectDocument.php
index 38090d8..e3dd790 100644
--- a/src/DataObject/DataObjectDocument.php
+++ b/src/DataObject/DataObjectDocument.php
@@ -343,8 +343,15 @@ class DataObjectDocument implements
      */
     public function toDBField(Field $field)
     {
+        $obj = $this->getDataObject();
         if ($field->getProperty()) {
-            return $this->getDataObject()->relObject($field->getProperty());
+            $firstField = explode('.', $field->getProperty())[0];
+            if ($obj->hasField($firstField) || $obj->hasMethod($firstField) || $obj->getRelationClass($firstField)) {
+                return $obj->relObject($field->getProperty());
+            } else {
+                // Guard against missing fields which only exist in some subclasses
+                return null;
+            }
         }
 
         return $this->resolveField($field->getSearchFieldName());
```
